### PR TITLE
Annotate metrics

### PR DESCRIPTION
**Pull request by prasunanand**
Changes made:
* Add mypy to setup.cfg
* Add ArrayLike type
* Modify ci scripts
* Add annotations to dask_ml.metrics

Partially fixes https://github.com/dask/dask-ml/issues/607

### CODE REVIEW

File: dask_ml/_typing.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+from pathlib import Path
+from typing import TypeVar
+
+import numpy as np
+
+from dask.array import Array
+
+if TYPE_CHECKING:
```
Comment (TomAugspurger) [inline dask_ml/_typing.py]: Shouldn't need this. Pandas needs it to avoid circular references.

File: dask_ml/_typing.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    from pandas.core.arrays.base import ExtensionArray  # noqa: F401
```
Comment (TomAugspurger) [inline dask_ml/_typing.py]: I suspect our algorithms need actual ndarrays or Dask arrays. They have methods like `.reshape` that aren't present on ExtensionArray. Let's be strict for now and add ExtensionArray as appropriate.

File: dask_ml/metrics/classification.py [hunk 3]
```diff
@@ -110,7 +122,7 @@ def log_loss(
 
     if y_pred.ndim > 1 and y_true.ndim == 1:
         y_true = y_true.reshape(-1, 1)
-        drop_axis = 1
+        drop_axis: Optional[int] = 1
```
Comment (TomAugspurger) [inline dask_ml/metrics/classification.py]: Is this optional?

File: dask_ml/metrics/pairwise.py [hunk 4]
```diff
@@ -54,8 +67,12 @@ def pairwise_distances(X, Y, metric="euclidean", n_jobs=None, **kwargs):
 
 
 def euclidean_distances(
-    X, Y=None, Y_norm_squared=None, squared=False, X_norm_squared=None
-):
+    X: ArrayLike,
+    Y: ArrayLike = None,
```
Comment (TomAugspurger) [inline dask_ml/metrics/pairwise.py]: Optional.

File: dask_ml/metrics/pairwise.py [hunk 5]
```diff
@@ -87,7 +104,9 @@ def euclidean_distances(
     return distances if squared else da.sqrt(distances)
 
 
-def check_pairwise_arrays(X, Y, precomputed=False):
+def check_pairwise_arrays(
+    X: ArrayLike, Y: ArrayLike, precomputed: bool = False
+) -> Tuple[ArrayLike, ArrayLike]:
```
Comment (TomAugspurger) [inline dask_ml/metrics/pairwise.py]: I don't know typing well enough, but I think there's a way to say that the `Type[X]` will exactly match the type of the first element of the result tuple? So you know that if you put in a dask.Array you'll get back a dask.Array rather than an numpy.ndarray?

File: dask_ml/metrics/regression.py [hunk 6]
```diff
@@ -1,15 +1,21 @@
...
 import dask.array as da
 import numpy as np
 import sklearn.metrics
 from dask.utils import derived_from
 
+from .._typing import ArrayLike
+
 
-def _check_sample_weight(sample_weight):
+def _check_sample_weight(sample_weight: Optional[str]):
```
Comment (TomAugspurger) [inline dask_ml/metrics/regression.py]: Should this be `Optional[ArrayLike]`?

File: dask_ml/metrics/regression.py [hunk 7]
```diff
@@ -24,8 +30,12 @@ def _check_reg_targets(y_true, y_pred, multioutput):
 
 @derived_from(sklearn.metrics)
 def mean_squared_error(
-    y_true, y_pred, sample_weight=None, multioutput="uniform_average", compute=True
-):
+    y_true: ArrayLike,
+    y_pred: ArrayLike,
+    sample_weight: Optional[ArrayLike] = None,
+    multioutput: Optional[str] = "uniform_average",
```
Comment (TomAugspurger) [inline dask_ml/metrics/regression.py]: Would be good to check with scikit-learn what the actual supported types are. Can this be an array, which just isn't implemented yet?

File: dask_ml/metrics/scorer.py [hunk 8]
```diff
@@ -17,7 +20,7 @@
 )
 
 
-def get_scorer(scoring, compute=True):
+def get_scorer(scoring: Union[str, Callable], compute=True) -> Callable:
```
Comment (TomAugspurger) [inline dask_ml/metrics/scorer.py]: Annotate compute.

**Comment by TomAugspurger** [comment]
If you merge master then the unrelated CI issues should be fixed.

File: dask_ml/metrics/classification.py (hunk 3 above)
Comment (prasunanand) [reply dask_ml/metrics/classification.py:125]: Yeah, it gets assingned as `None` on line 129

File: dask_ml/metrics/pairwise.py (hunk 5 above)
Comment (prasunanand) [reply dask_ml/metrics/pairwise.py:109]: I am not sure about this :/ .

File: dask_ml/metrics/regression.py (hunk 7 above)
Comment (prasunanand) [reply dask_ml/metrics/regression.py:36]: https://scikit-learn.org/stable/modules/generated/sklearn.metrics.mean_squared_error.html

`multi_output` is a string

**Comment by prasunanand** [comment]
@TomAugspurger , Request your review.

File: dask_ml/_typing.py [hunk 9]
```diff
@@ -0,0 +1,17 @@
+from pathlib import Path
```
Comment (TomAugspurger) [inline dask_ml/_typing.py:1]: Path seems to be unused?

File: dask_ml/_typing.py [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+from typing import TypeVar
+
+import numpy as np
+
+from dask.array import Array
+
+from pandas.core.arrays.base import ExtensionArray  # noqa: F401
```
Comment (TomAugspurger) [inline dask_ml/_typing.py:8]: Remove this one.

File: dask_ml/_typing.py [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+from pandas.core.indexes.base import Index  # noqa: F401
+from pandas.core.series import Series  # noqa: F401
+
+# array-like
+
+AnyArrayLike = TypeVar(
+    "AnyArrayLike", "Index", "Series", "Array", np.ndarray
+)
+ArrayLike = TypeVar("ArrayLike", "Array", np.ndarray)
```
Comment (TomAugspurger) [inline dask_ml/_typing.py:17]: Similarly, no quotes.

File: dask_ml/metrics/pairwise.py [hunk 12]
```diff
@@ -39,7 +46,13 @@ def pairwise_distances_argmin_min(
     return argmins, mins
 
 
-def pairwise_distances(X, Y, metric="euclidean", n_jobs=None, **kwargs):
+def pairwise_distances(
+    X: ArrayLike,
+    Y: ArrayLike,
+    metric: str = "euclidean",
```
Comment (TomAugspurger) [inline dask_ml/metrics/pairwise.py:52]: Same comment about `metric` accepting callable.

File: dask_ml/metrics/pairwise.py [hunk 13]
```diff
@@ -39,7 +46,13 @@ def pairwise_distances_argmin_min(
...
+    n_jobs: Optional[int] = None,
+    **kwargs
```
Comment (TomAugspurger) [inline dask_ml/metrics/pairwise.py:54]: Does mypy need anything special for `**kwargs`?

File: dask_ml/metrics/pairwise.py (hunk 5 above)
Comment (TomAugspurger) [reply dask_ml/metrics/pairwise.py:109]: OK to punt on it for now.

File: dask_ml/metrics/pairwise.py [hunk 14]
```diff
@@ -113,13 +132,15 @@ def check_pairwise_arrays(X, Y, precomputed=False):
...
-def linear_kernel(X, Y=None):
+def linear_kernel(X: ArrayLike, Y: Optional[ArrayLike] = None) -> ArrayLike:
     X, Y = check_pairwise_arrays(X, Y)
     return da.dot(X, Y.T)
 
 
 @derived_from(metrics.pairwise)
-def rbf_kernel(X, Y=None, gamma=None):
+def rbf_kernel(
+    X: ArrayLike, Y: Optional[ArrayLike] = None, gamma: Optional[int] = None
```
Comment (TomAugspurger) [inline dask_ml/metrics/pairwise.py:142]:
```suggestion
    X: ArrayLike, Y: Optional[ArrayLike] = None, gamma: Optional[float] = None
```

I think it's a float.

File: dask_ml/metrics/pairwise.py [hunk 15]
```diff
@@ -130,7 +151,13 @@ def rbf_kernel(X, Y=None, gamma=None):
 
 
 @derived_from(metrics.pairwise)
-def polynomial_kernel(X, Y=None, degree=3, gamma=None, coef0=1):
+def polynomial_kernel(
+    X: ArrayLike,
+    Y: Optional[ArrayLike] = None,
+    degree: int = 3,
+    gamma: Optional[int] = None,
```
Comment (TomAugspurger) [inline dask_ml/metrics/pairwise.py:158]:
```suggestion
    gamma: Optional[float] = None,
```

File: dask_ml/metrics/pairwise.py [hunk 16]
```diff
@@ -130,7 +151,13 @@ def rbf_kernel(X, Y=None, gamma=None):
...
+    coef0: int = 1,
```
Comment (TomAugspurger) [inline dask_ml/metrics/pairwise.py:159]:
```suggestion
    coef0: float = 1,
```

may need to change to 1.0?

File: dask_ml/metrics/pairwise.py [hunk 17]
```diff
@@ -140,7 +167,12 @@ def polynomial_kernel(X, Y=None, degree=3, gamma=None, coef0=1):
 
 
 @derived_from(metrics.pairwise)
-def sigmoid_kernel(X, Y=None, gamma=None, coef0=1):
+def sigmoid_kernel(
+    X: ArrayLike,
+    Y: Optional[ArrayLike] = None,
+    gamma: Optional[int] = None,
+    coef0: int = 1,
```
Comment (TomAugspurger) [inline dask_ml/metrics/pairwise.py:174]:
```suggestion
    gamma: Optional[float] = None,
    coef0: float = 1,
```

File: dask_ml/metrics/pairwise.py [hunk 18]
```diff
@@ -165,7 +197,14 @@ def sigmoid_kernel(X, Y=None, gamma=None, coef0=1):
 }
 
 
-def pairwise_kernels(X, Y=None, metric="linear", filter_params=False, n_jobs=1, **kwds):
+def pairwise_kernels(
+    X: ArrayLike,
+    Y: Optional[ArrayLike] = None,
+    metric: str = "linear",
```
Comment (TomAugspurger) [inline dask_ml/metrics/pairwise.py:203]: Can also be a callable.

File: dask_ml/metrics/pairwise.py [hunk 19]
```diff
@@ -165,7 +197,14 @@ def sigmoid_kernel(X, Y=None, gamma=None, coef0=1):
...
+    filter_params: bool = False,
+    n_jobs: int = 1,
```
Comment (TomAugspurger) [inline dask_ml/metrics/pairwise.py:205]: Optional, according to scikit-learn (they actually default to None. Not sure if that's a bug in dask-ml or deliberate).

**Comment by TomAugspurger** [comment]
Can you add `mypy` to the linting check in `azure-pipelines/posix.yaml`?

And there are some linting issues there which aren't failing the build. https://dev.azure.com/dask-dev/dask/_build/results?buildId=818&view=logs&j=34469798-9dec-5482-c96c-ad1eda9ec58d&t=3856b5af-eaa9-56ed-b4cc-e7b6fad51231&l=12 Can you fix those? I'll figure out why CI isn't failing.

**Comment by prasunanand** [comment]
Thanks @TomAugspurger . I am working on the changes.

File: dask_ml/metrics/pairwise.py (hunk 13 above)
Comment (prasunanand) [reply dask_ml/metrics/pairwise.py:54]: https://github.com/python/mypy/issues/5876

File: dask_ml/metrics/pairwise.py (hunk 19 above)
Comment (prasunanand) [reply dask_ml/metrics/pairwise.py:205]: Yeah. I think I should add `Optional[int]` for now.

**Comment by TomAugspurger** [comment]
Can you merge master? There will be a merge conflict with `posix.yaml`. You'll need to move your `mypy` check to `ci/code_checks.sh`.

**Comment by prasunanand** [comment]
I think it should have worked.
```
dask_ml/metrics/pairwise.py:218: error: Invalid index type "Union[str, Callable[[Any, Any], float]]" for "Dict[str, Any]"; expected type "str"
```
https://dev.azure.com/dask-dev/dask/_build/results?buildId=834&view=logs&j=9dcbdf5b-840d-5ddd-a5ac-58624932ec86&t=96578b6e-0c8a-5a4c-e6c4-a33b762a5e3a&l=25

**Comment by TomAugspurger** [comment]
> I think it should have worked.
cc @rth @thomasjpfan. IIRC you're investigating type annotations for scikit-learn.

**Comment by prasunanand** [comment]
Thanks :)
`assert isinstance(metric, str)`  worked

**Comment by prasunanand** [comment]
@TomAugspurger , It's all good now :)

**Comment by TomAugspurger** [comment]
Thanks, this looks good!

**Comment by rth** [comment]
> I'm not sure if there's a way for us to share types, since we have different definitions of "ArrayLike", but we can coordinate in places where we overlap.

@TomAugspurger  As far as I know, in scikit-learn we won't type array like objects for now precisely because it's unclear what the type for "ArrayLike" input should be. So at least there shouldn't be an issue of types being over-constrained on the scikit-learn side.

**Comment by mrocklin** [comment]
> So at least there shouldn't be an issue of types being over-constrained on the scikit-learn side.

This is true of most things I think :)
